nativelib/src/main/resources/stat.c

```diff
@@ -0,0 +1,168 @@
+#include <sys/stat.h>
+#include <gc.h>
+#include "types.h"
+
+// We don't use the "standard" types such as `dev_t` for instance
+// because these have different sizes on eg. Linux and OSX. We use the
+// smallest type that can hold all the possible values for the different systems.
+struct scalanative_stat {
+    scalanative_dev_t st_dev;         /** Device ID of device containing file. */
+    scalanative_dev_t st_rdev;        /** Device ID (if file is character or block special). */
+    scalanative_ino_t st_ino;         /** File serial number. */
+    scalanative_uid_t st_uid;         /** User ID of file. */
+    scalanative_gid_t st_gid;         /** Group ID of file. */
+    scalanative_off_t st_size;        /** For regular files, the file size in bytes.
+			                  For symbolic links, the length in bytes of the
+			                  pathname contained in the symbolic link.
+			                  For a shared memory object, the length in bytes.
+			                  For a typed memory object, the length in bytes.
+			                  For other file types, the use of this field is
+			                  unspecified. */
+    scalanative_time_t _st_atime;     /** Time of last access. */
+    scalanative_time_t _st_mtime;     /** Time of last data modification. */
+    scalanative_time_t _st_ctime;     /** Time of last status change. */
+    scalanative_blkcnt_t st_blocks;   /** Number of blocks allocated for this object. */
+    scalanative_blksize_t st_blksize; /** A file system-specific preferred I/O block size for
+			                  this object. In some file system types, this may
+			                  vary from file to file. */
+    scalanative_nlink_t st_nlink;     /** Number of hard links to the file. */
+    scalanative_mode_t st_mode;       /** Mode of file (see below). */
+};
+
+void scalanative_stat_init(struct stat *stat, struct scalanative_stat * my_stat) {
+    my_stat->st_dev     = stat->st_dev;
+    my_stat->st_rdev    = stat->st_rdev;
+    my_stat->st_ino     = stat->st_ino;
+    my_stat->st_uid     = stat->st_uid;
+    my_stat->st_gid     = stat->st_gid;
+    my_stat->st_size    = stat->st_size;
+    my_stat->_st_atime  = stat->st_atime;
+    my_stat->_st_mtime  = stat->st_mtime;
+    my_stat->_st_ctime  = stat->st_ctime;
+    my_stat->st_blksize = stat->st_blksize;
+    my_stat->st_blocks  = stat->st_blocks;
+    my_stat->st_nlink   = stat->st_nlink;
+    my_stat->st_mode    = stat->st_mode;
+}
+
+struct scalanative_stat *scalanative_stat_copy(struct stat *stat) {
+    struct scalanative_stat *my_stat =
+        (struct scalanative_stat *) GC_malloc(sizeof(struct scalanative_stat));
```

Comment:
Bindings and wrappers around standard C APIs should *never* GC allocate. In fact, unmanaged pointers (i.e. `Ptr[_]`) to the GC heap are completely implementation-specific and should never be used anywhere outside `scalanative.runtime`. Instead, the standard way to do it in C way is to take an extra "out" argument where you write the result. Then the caller has to take care of allocation and deallocation, the way he likes.